### PR TITLE
Allow applications and scripts to be launched within the sessions

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
A new `--script` and `--app` flags have been added to launch `scripts` and `applications` within the sessions. `scripts` are intended to be launched within a new terminal, where applications are launched directly. In ever case the `DISPLAY` variable is set, so both flags accept scripts.

Due to the idiosyncrasies in the various session `types`, the two flags shell out to `launch-script` and `launch-app` shims respectively. This allows session type specific setup and a way to disable the flags per type.

Multiple apps can be launched be repeating the `--app` flag (similarly for `--script`). This required upgrading `commander` to version `2.2`

### CODE REVIEW

**Review by benarmston — CHANGES_REQUESTED** [review]
This works well for both `gnome` and `xfce`, but I think both `terminal` and `xterm` need some work.

The problem with using `screen` is that its session hangs around after the desktop session has terminated and potentially after the interactive job has timedout.  We don't want to rely on users cleaning up after themselves.

Another issue with using `screen` is that the user still needs to connect to the screen session after the VNC session has started.  It would be a much nicer experience if they were immediately presented with the programs they have launched.  This could be achieved by having the `terminal/session.sh` script run the script it is given. 

This would result in terminal sessions only supporting a single script, but in practice that is all that they are going to need.

File: lib/desktop/cli.rb [hunk 1]
```diff
@@ -172,6 +160,18 @@ def cli_syntax(command, args_str = nil)
       c.summary = 'Start an interactive desktop session'
       c.action Commands, :start
       c.option '-g', '--geometry GEOMETRY', Geometry, 'Specify desktop geometry.'
+      c.slop.array '-a', '--app', <<~DESC.chomp, delimiter: nil, meta: '"BINARY [ARGUMENTS...]"'
+        Specify the binary/launch script for a graphical application.
+
+        Multiple applications can be started by repating the --app flag.
+        Not supported by all desktop types.
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/desktop/cli.rb]: Updated

File: lib/desktop/cli.rb [hunk 2]
```diff
@@ -172,6 +160,18 @@ def cli_syntax(command, args_str = nil)
...
+      DESC
+      c.slop.array '-s', '--script', <<~DESC.chomp, delimiter: nil, meta: '"SCRIPT [ARGUMENTS...]"'
+        Start a script on the remote session.
+
+        Mutliple scripts can be started by repeating the --script flag.
+        Not supported by all desktop types.
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply lib/desktop/cli.rb]: Updated

**Comment by WilliamMcCumstie** [comment]
The launching of scripts for `xterm` and `terminal` has been reworked. There are now three mechanisms to launch app/scripts:

1. The `--app` will call `launch-app.sh` to create graphical applications,
2. The `--script` will attempt to call `launch-script.sh` to create a graphical terminal, OR
3. `--script` will fallback on patching the `session.sh` to start a single script.

Options 1/2 are the same as above and are used by `xfce` and `gnome`. The `--app` and `--script` flags can be repeated to launch multiple applications.

However `xterm` and `terminal` now flag that they are `scriptable` within their metadata. This denotes they can run a script via their `session.sh`.

Because `flight-desktop` is built on `vncserver`, it can only specify a single `-sessionscript`. This means the `session.sh` must contain the desktop specific start code as well as the script command. This has been achieved by patching the `script` into `session.sh` when it is "installed".

**Comment by benarmston** [comment]
Rebeased on `master`.

**Pull request merged**